### PR TITLE
Parse unnamed struct and union fields

### PR DESCRIPTION
**Pull request by carbotaniuman**
Reincorporates code from https://github.com/rust-lang/rust/pull/84571 and https://github.com/rust-lang/rust/pull/85515,  for #49804 which was reverted due to context-sensitive issues around the `union` keyword.

This also incorporates a modified version of the error recovery from https://github.com/rust-lang/rust/pull/88815, with the modifications being in where the parser recovery occurs.

The main issue I'm facing right now is in recovering for these types, as the current implementation fails to properly reject `impl union {} for union {} {}` without also incorrectly rejecting `impl union {}` and I'm not really sure how to resolve that without making the code a mess.

### CODE REVIEW

File: src/tools/rustfmt/src/types.rs [hunk 1]
```diff
@@ -799,6 +799,54 @@ impl Rewrite for ast::Ty {
...
+                    is_placeholder: false,
+                };
+                format_struct_struct(
+                    &context,
+                    &StructParts::from_variant(&variant),
+                    fields,
+                    shape.indent,
+                    None,
+                )
+            }
```
Comment (calebcartwright) [inline src/tools/rustfmt/src/types.rs:849]: For these two new variants we should just emit whatever was originally in the span.

That can be accomplished via `Some(context.snippet(self.span).to_owned())` (similar to the `Err` variant)

**Review by calebcartwright — CHANGES_REQUESTED** [review]
Review only applicable to the rustfmt subset of the diff

Thanks for including an implementation, but it's actually not necessary to do so and generally our preference in these new cases to leave the formatting untouched initially. We'll probably end up with something fairly similar on the rustfmt side, but we'll take care of that a little later on.

Inline comment left below on the one change requested for rustfmt, then believe the rustfmt changes in the other files could be reverted

**Comment by compiler-errors** [comment]
Sorry for taking so long to get to this, not sure if I have enough context to review this.

r? compiler

File: compiler/rustc_feature/src/active.rs [hunk 2]
```diff
@@ -521,6 +521,8 @@ declare_features! (
     /// Allows creation of instances of a struct by moving fields that have
     /// not changed from prior instances of the same struct (RFC #2528)
     (active, type_changing_struct_update, "1.58.0", Some(86555), None),
+    /// Allows unnamed fields of struct and union type
+    (incomplete, unnamed_fields, "1.53.0", Some(49804), None),
```
Comment (fee1-dead) [inline compiler/rustc_feature/src/active.rs:525]: The version number does not seem to be right..

File: src/test/ui/parser/issues/issue-88583-union-as-ident.rs [hunk 3]
```diff
@@ -6,7 +6,8 @@ struct union;
 
 impl union {
     pub fn new() -> Self {
-        union { }
+        let a = union {};
+        a
```
Comment (fee1-dead) [inline src/test/ui/parser/issues/issue-88583-union-as-ident.rs:10]: Is this change necessary? Does it still pass without this change?

**Comment by fee1-dead** [comment]
I am not very familiar with ast code.

r? compiler

**Comment by cjgillot** [comment]
If I understand the RFC correctly, we are trying to allow unnamed fields in struct & unions with either named or unnamed struct/union type.  So an unnamed struct/union can only happen inside a `_` field.  You add unnamed struct & unions as type kinds.  I'm not sure this is the best direction.
What about changing `FieldDef` using:
```rust
enum FieldDefKind {
    Named(Ident, P<Ty>),
    Unnamed(P<Ty>),
    Nested(StructOrUnion, Vec<FieldDef>)
}
```

This would probably remove a lot of the headaches about parsing ambiguities and having `union {}` pop up in unexpected places.
What do you think?

**Comment by carbotaniuman** [comment]
That approach is #84415, and a comment was made suggesting the other approach. https://github.com/rust-lang/rust/pull/84415#issuecomment-826155552

Is the reasoning expressed in the comment still valid?

File: compiler/rustc_ast/src/visit.rs [hunk 4]
```diff
@@ -456,6 +456,9 @@ pub fn walk_ty<'a, V: Visitor<'a>>(visitor: &mut V, typ: &'a Ty) {
         TyKind::Typeof(ref expression) => visitor.visit_anon_const(expression),
         TyKind::Infer | TyKind::ImplicitSelf | TyKind::Err => {}
         TyKind::MacCall(ref mac) => visitor.visit_mac_call(mac),
+        TyKind::AnonymousStruct(ref fields, ..) | TyKind::AnonymousUnion(ref fields, ..) => {
```
Comment (cjgillot) [inline compiler/rustc_ast/src/visit.rs:459]: Please explicitly ignore fields in visitors, so we have a compiler error when we change them.
```suggestion
        TyKind::AnonymousStruct(ref fields, _recovered) | TyKind::AnonymousUnion(ref fields, _recovered) => {
```

File: compiler/rustc_ast/src/mut_visit.rs [hunk 5]
```diff
@@ -489,6 +489,9 @@ pub fn noop_visit_ty<T: MutVisitor>(ty: &mut P<Ty>, vis: &mut T) {
             visit_vec(bounds, |bound| vis.visit_param_bound(bound));
         }
         TyKind::MacCall(mac) => vis.visit_mac_call(mac),
+        TyKind::AnonymousStruct(fields, ..) | TyKind::AnonymousUnion(fields, ..) => {
```
Comment (cjgillot) [inline compiler/rustc_ast/src/mut_visit.rs:492]: Please explicitly ignore fields in visitors, so we have a compiler error when we change them.
```suggestion
        TyKind::AnonymousStruct(fields, _recovered) | TyKind::AnonymousUnion(fields, _recovered) => {
```

File: compiler/rustc_ast/src/ast.rs [hunk 6]
```diff
@@ -2021,6 +2021,10 @@ pub enum TyKind {
     Never,
     /// A tuple (`(A, B, C, D,...)`).
     Tup(Vec<P<Ty>>),
+    /// An anonymous struct type i.e. `struct { foo: Type }`
+    AnonymousStruct(Vec<FieldDef>, bool),
```
Comment (cjgillot) [inline compiler/rustc_ast/src/ast.rs:2025]: Could you add a comment saying what is the `bool`?  Where do we need it?

File: compiler/rustc_ast_lowering/src/item.rs [hunk 7]
```diff
@@ -723,7 +723,10 @@ impl<'hir> LoweringContext<'_, 'hir> {
         }
     }
 
-    fn lower_field_def(&mut self, (index, f): (usize, &FieldDef)) -> hir::FieldDef<'hir> {
+    pub(super) fn lower_field_def(
```
Comment (cjgillot) [inline compiler/rustc_ast_lowering/src/item.rs:726]: Stray change?

File: compiler/rustc_parse/src/parser/item.rs [hunk 8]
```diff
@@ -550,52 +552,103 @@ impl<'a> Parser<'a> {
...
             None
         };
 
         generics.where_clause = self.parse_where_clause()?;
 
-        let impl_items = self.parse_item_list(attrs, |p| p.parse_impl_item(ForceCollect::No))?;
+        let (mut impl_items, err) =
+            self.parse_item_list(attrs, |p| p.parse_impl_item(ForceCollect::No))?;
+
+        if let Some(mut err) = err {
```
Comment (cjgillot) [inline compiler/rustc_parse/src/parser/item.rs:605]: Could you add a comment explaining what happens here?

File: compiler/rustc_parse/src/parser/ty.rs [hunk 9]
```diff
@@ -326,6 +410,25 @@ impl<'a> Parser<'a> {
         if allow_qpath_recovery { self.maybe_recover_from_bad_qpath(ty) } else { Ok(ty) }
     }
 
+    pub(crate) fn parse_anonymous_ty(&mut self, lo: Span) -> PResult<'a, TyKind> {
+        let is_union = self.token.is_keyword(kw::Union);
```
Comment (cjgillot) [inline compiler/rustc_parse/src/parser/ty.rs:414]: Could you assert that the keyword is `Struct` if not `Union`?

**Review by cjgillot — COMMENTED** [review]
Thanks for the explanation @carbotaniuman. I left a few questions.

File: compiler/rustc_parse/src/parser/ty.rs [hunk 10]
```diff
@@ -284,14 +328,35 @@ impl<'a> Parser<'a> {
...
+                    Ok(ty) => ty,
+                    Err(err) => {
+                        err.cancel();
+                        self.restore_snapshot(snapshot);
+                        self.parse_path_start_ty(lo, allow_plus, ty_generics)?
+                    }
+                }
+            } else {
+                self.parse_path_start_ty(lo, allow_plus, ty_generics)?
+            }
```
[the remark this answers is not in the document]
Comment (carbotaniuman) [reply compiler/rustc_parse/src/parser/ty.rs:359]: Ok, so this is the part of the change that I'm least sure of.  In this code, because `union` can be an identifier, you have to expressly opt-in to parsing an anonymous union type. `struct` on the other hand is a keyword and will attempt to parse an anonymous struct type for recovery. This is really more of hack than anything proper.

This has some implications, such as `impl struct {}` having an error mentioning anonymous struct types (instead of expected type found keyword), but `impl struct{} {}` having a "proper" error. This may require recovery changes in the impl block parsing beyond what I have already done, but I'm not too sure on this.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@carbotaniuman - can you please post your status on this PR? Thank you

**Comment by carbotaniuman** [comment]
I've been busy with other stuff, but I'm back to working on this.

File: compiler/rustc_ast_lowering/src/item.rs (hunk 7 above)
Comment (carbotaniuman) [reply compiler/rustc_ast_lowering/src/item.rs:752]: If use the function somewhere else outside of this module.

File: compiler/rustc_parse/src/parser/item.rs (hunk 8 above)
Comment (carbotaniuman) [reply compiler/rustc_parse/src/parser/item.rs:615]: Uhh this is with the other comment that I'm not too sure if this is the right way to do things, but basically I'm trying to parse an impl block, if that fails I try and parse an anonymous struct or union type and then parse an impl block. (Really I would need to do it three times to recover form `impl union {}`, but again I'm not even sure this is the right path.

Comment (carbotaniuman) [reply compiler/rustc_parse/src/parser/item.rs:615]: It's less that error message is bad and it's more that I can't do recovery (at least how I'm doing it) because calling the method during recovery will emit the error without allowing me to suppress it.

**Comment by apiraino** [comment]
Switching review status as at a glance I'm unsure if the latest commits replied in full to the [previous review](url). Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

**Comment by carbotaniuman** [comment]
@rustbot ready 

Mainly want a review on the general approach for error handling here, as I'm not too sure it's correct.

**Comment by WaffleLapkin** [comment]
Marking as waiting on author again (merge conflicts and failing CI)

**Comment by tmandry** [comment]
@carbotaniuman are you still working on this?

@cjgillot would you be willing to provide feedback on the error handling questions around https://github.com/rust-lang/rust/pull/99754#pullrequestreview-1072051308?

File: compiler/rustc_parse/src/parser/ty.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (joshtriplett) [reply compiler/rustc_parse/src/parser/ty.rs:389]: @carbotaniuman:

@Amanieu and I were reviewing this, and wondering if it would be possible to confine this *entirely* to field definitions, and even then only when the field name is `_`.

If the field name is `_`, then look for `struct` or `union` in the type. If the type is `struct` then it *must* continue on with `{`. If the type is `union`, it *could* be `union;` or `union::something...`, either of which would be the *other* case of unnamed fields (where you have a named type), but if it isn't the start of a path then it has to continue with `{`.

Either way, we're wondering if this can *only* be parsed in field types, and nowhere else, avoiding the need to handle it (for recovery or otherwise) in other locations a type can appear.

**Comment by carbotaniuman** [comment]
I've been on vacation, but I'm able to respond and push the work through
now.

All of the parser recovery here is to pass the tests that make recovery
happen, and not just error out if an unnamed type is present somewhere
else. I can of course not have this recovery and simplify the code, but I'm
not sure if that's something I can do?

**Comment by joshtriplett** [comment]
@carbotaniuman Ah, I see. So, by parsing the entire type, the parser then knows how much to skip over after saying that unnamed types aren't allowed in the context (rather than just encountering `struct` or `union` and failing)?

That makes sense.

**Comment by JohnCSimon** [comment]
@carbotaniuman 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
